packages/pg-cursor/index.js

```diff
@@ -6,230 +6,232 @@ const util = require('util')
+  handlePortalSuspended() {
+    this._sendRows()
   }
-}
 
-Cursor.prototype._getRows = function (rows, cb) {
-  this.state = 'busy'
-  this._cb = cb
-  this._rows = []
-  const msg = {
-    portal: this._portal,
-    rows: rows,
+  handleReadyForQuery() {
+    this._sendRows()
+    this.state = 'done'
+    this.emit('end', this._result)
   }
-  this.connection.execute(msg, true)
-  this.connection.flush()
-}
 
-// users really shouldn't be calling 'end' here and terminating a connection to postgres
-// via the low level connection.end api
-Cursor.prototype.end = util.deprecate(function (cb) {
-  if (this.state !== 'initialized') {
+  handleEmptyQuery() {
     this.connection.sync()
   }
-  this.connection.once('end', cb)
-  this.connection.end()
-}, 'Cursor.end is deprecated. Call end on the client itself to end a connection to the database.')
 
-Cursor.prototype.close = function (cb) {
-  if (!this.connection || this.state === 'done') {
-    if (cb) {
-      return setImmediate(cb)
-    } else {
-      return
+  handleError(msg) {
+    // If we're in an initialized state we've never been submitted
+    // and don't have a connection instance reference yet.
+    // This can happen if you queue a stream and close the client before
+    // the client has submitted the stream.  In this scenario we don't have
+    // a connection so there's nothing to unsubscribe from.
+    if (this.state !== 'initialized') {
+      this.connection.removeListener('noData', this._ifNoData)
+      this.connection.removeListener('rowDescription', this._rowDescription)
+      // call sync to trigger a readyForQuery
+      this.connection.sync()
     }
-  }
 
-  this._closePortal()
-  this.state = 'done'
-  if (cb) {
-    this.connection.once('readyForQuery', function () {
-      cb()
-    })
-  }
-}
+    this.state = 'error'
+    this._error = msg
+    // satisfy any waiting callback
+    if (this._cb) {
+      this._cb(msg)
+    }
+    // dispatch error to all waiting callbacks
+    for (let i = 0; i < this._queue.length; i++) {
+      this._queue.pop()[1](msg)
+    }
 
-Cursor.prototype.read = function (rows, cb) {
-  if (this.state === 'idle' || this.state === 'submitted') {
-    return this._getRows(rows, cb)
-  }
-  if (this.state === 'busy' || this.state === 'initialized') {
-    return this._queue.push([rows, cb])
+    if (this.listenerCount('error') > 0) {
+      // only dispatch error events if we have a listener
+      this.emit('error', msg)
+    }
   }
-  if (this.state === 'error') {
-    return setImmediate(() => cb(this._error))
+
+  _getRows(rows, cb) {
+    this.state = 'busy'
+    this._cb = cb
+    this._rows = []
+    const msg = {
+      portal: this._portal,
+      rows: rows,
+    }
+    this.connection.execute(msg, true)
+    this.connection.flush()
+  }
+
+  // users really shouldn't be calling 'end' here and terminating a connection to postgres
+  // via the low level connection.end api
+  end(cb) {
+    return util.deprecate(function () {
```

Comment:
This will print a deprecation warning every time it’s called. How about, after the class,

```js
Cursor.prototype.end = util.deprecate(Cursor.prototype.end, …)
```

Reply:
Yeah I thought about that, but it's pretty bad practice to mix. I was also thinking of putting it in the constructor, but I think that might be worse?

Reply:
Yes, putting it in the constructor would be wrong. The line I suggested isn’t bad practice.

Reply:
Not in of itself, but mixing it. I suppose that does seem to be the best option all things considered. I'll change that tomorrow.